### PR TITLE
New play to change bican routing between CHN/CMN

### PR DESCRIPTION
**Pull request by bklei**
## Summary and Scope

DO NOT MERGE YET -- SOLICITING FEEDBACK/TESTING

## Issues and Related PRs

* Resolves [CASMNET-829](https://jira-pro.its.hpecorp.net:8443/browse/CASMNET-829)

## Testing

```
ncn-m001:~ # kubectl logs -n services cfs-3c35a396-943a-4486-b0b0-3cbd5de3e6a4-zns7f ansible-0 -f
Waiting for Inventory
Waiting for Inventory
Inventory generation completed
SSH keys migrated to /root/.ssh
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
HTTP/1.1 200 OK
content-type: text/html; charset=UTF-8
cache-control: no-cache, max-age=0
x-content-type-options: nosniff
date: Thu, 17 Feb 2022 22:28:20 GMT
server: envoy
transfer-encoding: chunked

Sidecar available
Using /etc/ansible/ansible.cfg as config file

PLAY [Management] **************************************************************

TASK [csm.ncn.move_bican_routing : Get the bican toggle from SLS] **************
changed: [x3000c0s7b0n0] => {"changed": true, "cmd": "/opt/cray/bican-utils/bi-can-get-sls-toggle.py", "delta": "0:00:03.076657", "end": "2022-02-17 22:28:25.893555", "rc": 0, "start": "2022-02-17 22:28:22.816898", "stderr": "", "stderr_lines": [], "stdout": "CMN", "stdout_lines": ["CMN"]}

TASK [csm.ncn.move_bican_routing : Setup the interface for either CMN/CHN] *****
changed: [x3000c0s7b0n0] => {"changed": true, "cmd": "/opt/cray/bican-utils/bi-can-interface.py CMN", "delta": "0:00:03.294673", "end": "2022-02-17 22:28:29.807624", "rc": 0, "start": "2022-02-17 22:28:26.512951", "stderr": "", "stderr_lines": [], "stdout": "hostname: ncn-w004\nCMN IP from SLS: 10.103.0.22\nCMN_Network from SLS: 10.103.0.0/25\nCMN_VLAN from SLS: 7\n/etc/sysconfig/network/ifcfg-bond0.cmn0 exists\nCMN interface already configured\n('INTERFACE: bond0.cmn0 STATE: up CARRIER: up IP: 10.103.0.22', '10.103.0.22')\nPing test to 10.103.0.1 from the CMN interface is successful", "stdout_lines": ["hostname: ncn-w004", "CMN IP from SLS: 10.103.0.22", "CMN_Network from SLS: 10.103.0.0/25", "CMN_VLAN from SLS: 7", "/etc/sysconfig/network/ifcfg-bond0.cmn0 exists", "CMN interface already configured", "('INTERFACE: bond0.cmn0 STATE: up CARRIER: up IP: 10.103.0.22', '10.103.0.22')", "Ping test to 10.103.0.1 from the CMN interface is successful"]}

TASK [csm.ncn.move_bican_routing : Set the default route for either CMN/CHN] ***
changed: [x3000c0s7b0n0] => {"changed": true, "cmd": "/opt/cray/bican-utils/bi-can-default-route.py CMN", "delta": "0:00:03.169459", "end": "2022-02-17 22:28:33.597849", "rc": 0, "start": "2022-02-17 22:28:30.428390", "stderr": "", "stderr_lines": [], "stdout": "CMN Default Gateway 10.103.0.1\nThe current default gateway is 10.103.0.1\nINTERFACE STATUS\n['bond0.cmn0']\n('INTERFACE: bond0.cmn0 STATE: up CARRIER: up IP: 10.103.0.22', '10.103.0.22')\n\nyour default route is already set to 10.103.0.1\n/etc/sysconfig/network/ifroute-bond0.cmn0 already exists and file contents are correct", "stdout_lines": ["CMN Default Gateway 10.103.0.1", "The current default gateway is 10.103.0.1", "INTERFACE STATUS", "['bond0.cmn0']", "('INTERFACE: bond0.cmn0 STATE: up CARRIER: up IP: 10.103.0.22', '10.103.0.22')", "", "your default route is already set to 10.103.0.1", "/etc/sysconfig/network/ifroute-bond0.cmn0 already exists and file contents are correct"]}

TASK [csm.ncn.move_bican_routing : Enable SSH for either CMN/CHN] **************
changed: [x3000c0s7b0n0] => {"changed": true, "cmd": "/opt/cray/bican-utils/bi-can-ssh.py CMN", "delta": "0:00:01.282049", "end": "2022-02-17 22:28:35.488233", "rc": 0, "start": "2022-02-17 22:28:34.206184", "stderr": "", "stderr_lines": [], "stdout": "CMN interface configured\n10.103.0.22\nINTERFACE: bond0.cmn0 STATE: up CARRIER: up IP: 10.103.0.22\n\"ListenAddress 10.103.0.22\" is in /etc/ssh/sshd_config\nThis is the correct configuration", "stdout_lines": ["CMN interface configured", "10.103.0.22", "INTERFACE: bond0.cmn0 STATE: up CARRIER: up IP: 10.103.0.22", "\"ListenAddress 10.103.0.22\" is in /etc/ssh/sshd_config", "This is the correct configuration"]}

PLAY RECAP *********************************************************************
x3000c0s7b0n0              : ok=4    changed=4    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

CFS Status Aggregator encountered a problem. The Ansible play has completed, and the following errors impact only status reporting.
These warnings can be ignored if the component is not expected to be managed by CFS and is not expected in the CFS components database.
CFS Status Aggregator: Unable to find commit hash
[WARNING]: Failure using method (v2_playbook_on_stats) in callback plugin
(<ansible.plugins.callback.cfs_aggregator.CallbackModule object at
0x7f1d857c7860>): [Errno 2] No such file or directory:
'/inventory/.git/refs/heads/CASMNET-829-1.2'
```

### Tested on:

  * `hela`

### Test description:

Ran play against worker on hela, exercised the code as far as it would go (created interface/route files, just can't yet ping the CHN gateway for default route switch)

## Risks and Mitigations

Low

## Pull Request Checklist

- [ ] Version number(s) incremented, if applicable
- [ ] Copyrights updated
- [ ] License file intact
- [ ] Target branch correct
- [ ] CHANGELOG.md updated
- [ ] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

File: ansible/roles/csm.ncn.move_bican_routing/files/bi-can-default-route.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# (MIT License)
+import pprint
+import os
+from pyroute2 import IPRoute
```
Comment (rkleinman-hpe) [inline ansible/roles/csm.ncn.move_bican_routing/files/bi-can-default-route.py:25]: Is this available in the AEE image?

**Review by rkleinman-hpe — COMMENTED** [review]
Quite a lot of python added to this role. Would these script be better served bundled as an RPM and installed on the NCNs?

File: ansible/roles/csm.ncn.move_bican_routing/files/bi-can-default-route.py (hunk 1 above)
Comment (trad511) [reply ansible/roles/csm.ncn.move_bican_routing/files/bi-can-default-route.py:25]: It's been in the NCN images for a few months now.

**Review by rkleinman-hpe — CHANGES_REQUESTED** [review]
Please increment the `.y` file in the root of the repo. This looks like it is adding a backwards-compatible feature.
(brad): thx for the tip -- fixed

File: ansible/roles/csm.ncn.move_bican_routing/files/bi-can-default-route.py (hunk 1 above)
Comment (bklei) [reply ansible/roles/csm.ncn.move_bican_routing/files/bi-can-default-route.py]: After further discussion with sean -- we decided to go the rpm route (https://github.com/Cray-HPE/casmnet-bican/pull/2) -- have removed all the python from this play...

**Comment by mharding-hpe** [comment]
@bklei can this PR be closed? Or is it still something you are considering?

**Comment by mharding-hpe** [comment]
I'm gonna close this since it hasn't been touched in over a year. It can be reopened if it is still needed.

**Pull request closed without merging**